src/script/interpreter.cpp

```diff
@@ -2092,6 +2091,105 @@ bool VerifyTaprootCommitment(uint256 &tapleaf_hash,
+        }
+        if (stack.size() == 1) {
+            // Spend using single signature instead of executing script
+            valtype &vch_sig = stacktop(-1);
+            bool f_success = false;
+            if (!EvalChecksig(vch_sig, vch_pubkey, script_pubkey.begin(),
+                              script_pubkey.end(),
+                              flags | SCRIPT_REQUIRE_TAPROOT_SIGHASH, checker,
+                              metrics_out, std::nullopt, serror, f_success)) {
+                if (*serror == ScriptError::SIG_NULLFAIL) {
+                    *serror = ScriptError::TAPROOT_VERIFY_SIGNATURE_FAILED;
+                }
+                // serror is set
+                return false;
+            }
+            if (!f_success) {
+                return set_error(serror,
+                                 ScriptError::TAPROOT_VERIFY_SIGNATURE_FAILED);
+            }
+            return set_success(serror);
+        } else {
+            // Spend using executing script, internal pubkey and merkle path
+            valtype control_block = stacktop(-1);
+            valtype script_bytes = stacktop(-2);
+            CScript exec_script(script_bytes.begin(), script_bytes.end());
+            popstack(stack);
+            popstack(stack);
+            const uint32_t size_remainder =
+                (control_block.size() - TAPROOT_CONTROL_BASE_SIZE) %
+                TAPROOT_CONTROL_NODE_SIZE;
+            if (control_block.size() < TAPROOT_CONTROL_BASE_SIZE ||
+                control_block.size() > TAPROOT_CONTROL_MAX_SIZE ||
+                (size_remainder != 0)) {
+                return set_error(serror,
+                                 ScriptError::TAPROOT_WRONG_CONTROL_SIZE);
+            }
+            if ((control_block[0] & TAPROOT_LEAF_MASK) !=
+                TAPROOT_LEAF_TAPSCRIPT) {
+                return set_error(
+                    serror, ScriptError::TAPROOT_LEAF_VERSION_NOT_SUPPORTED);
+            }
+            uint256 tapleaf_hash;
+            if (!VerifyTaprootCommitment(tapleaf_hash, control_block,
+                                         vch_pubkey, exec_script)) {
+                return set_error(serror,
+                                 ScriptError::TAPROOT_VERIFY_COMMITMENT_FAILED);
+            }
+            if (script_pubkey.size() == SCRIPT_SIZE_WITH_STATE) {
+                stack.push_back(valtype(
+                    script_pubkey.begin() + SCRIPT_SIZE_WITHOUT_STATE + 1,
+                    script_pubkey.begin() + SCRIPT_SIZE_WITH_STATE));
+            }
+            ScriptExecutionData execdata{tapleaf_hash};
+            if (!EvalScript(stack, exec_script, flags, checker, metrics_out,
+                            execdata, serror)) {
+                // serror is set
+                return false;
+            }
+        }
+    } else {
+        return set_error(serror, ScriptError::SCRIPTTYPE_INVALID_TYPE);
+    }
+    if (stack.empty() || !CastToBool(stack.back())) {
```

Comment:
In order to get here we need `script_pubkey[1] == OP_1` and a failure to return somewhere above in verifying a taproot script.  This seems like a logic error.

Reply:
This happens when there is success in evaluating the scripts above. Then we're essentially checking if the script passed?

Reply:
Doesn't this not ensure cleanstack?